### ISSUE TITLE
alice private key,HD chain issue?

### ISSUE THREAD

**Issue by sidhujag** [omnilaboratory/OmniBOLT-spec#1]
what temporary key is Alice's used to send to bob in your examples? Most wallets now are HD key based and so when an HD key is derived it will send that private key, so potentially any private keys can be found at the account level of Alice's hd chain if Bob finds out about Alice's pubkey one or more levels up from accounts in the bip32 strategy. Am I thinking about it the wrong way? Where is this key created I assume it would be an address where Alice owns some asset in which case it would be a key in the hd keychain likely.

**Comment by neocarmack** [comment]
I guess the context we are talking about is when Alice pay Bob some USDT in the channel, she has to send bob the private key <Alice2> of the 2-2 multisig address <Alice2, Bob>, where gets the 60 USDT in our diagram in the chapter 2, commitment_tx.

We leverage HD to generate public/private key pair for any address created in any process of payment. Alice has a wallet, on her desktop or her phone, where she stores her mnemonic words, which is the root that generats all the derivated key pairs. The mnemonic words are the root of the HD tree, and can not be leak to anyone else. Leak of a single pair of public/private key will not jeopardize other information generated by the same set of mnemonic words.

So as long as the mnemonic words are safe and unknown to others, the tree of public/private key pairs are safe. No one can use a private key to reversely decode the mneminic words.

**Comment by sidhujag** [comment]
I don’t think you understand how it works.

A xpub in an HD chain that is not hardened along with any child private key can derive all grandchildren keys.

“ Note however that the following properties does not exist:
Given a parent extended public key (Kpar,cpar) and a child public key (Ki), it is hard to find i.
Given a parent extended public key (Kpar,cpar) and a non-hardened child private key (ki), it is hard to find kpar.”

From bip32. 

Why not just use the r preimage that ln holy suggests. Why use private key in the first place.

**Comment by neocarmack** [comment]
Correct. There is no room for non-hardened keys to be applied, especially the temporary private key <Alice 2>, which is hardened  and will be hand over to Bob during the next commitment transaction.

An alternative way to keep safe is to apply a set of complete new mnemonic words to generate the temporary key pairs, and only used for once. But this will bring extra complexity in implementation.

In creating HTLC, the private key we use to replace R, whose preimage is used to lock a HTLC, is hardened ofcourse, otherwise there is a huge risk of leaking private keys of the account.

**Comment by neocarmack** [comment]
According to BIP44, which uses hardening/private derivation on most levels: 
m / purpose' / coin_type' / account' / change / address_index

but not on the change and address index level.

So a better practice is to apply temporary generated seed for every temporary multi-sig address. They dont live in the same HD keys chain. Then there is no chance to compromise any parent private keys.

**Comment by sidhujag** [comment]
> Correct. There is no place for non-hardened keys to be applied, especially the temporary private key <Alice 2>, which is hardened and will be handed over to Bob during the next commitment transaction.
> 
> An alternative way to keep safe is to apply a set of complete new mnemonic words to generate the temporary key pairs, and only use for once. But this will bring extra complexity in implementation.
> 
> In creating HTLC, the private key we use to replace preimage R, whose hashed value is used to lock a HTLC, is hardened ofcourse, otherwise there is a huge risk of leaking private keys of the account.

Sure. but why go through the hoops when you can simply just use a secret, why use the private key? If you are using a new HD chain now you have to manage that new HD chain and ensure its a new chain for every transaction. Why not just use a normal secret?

**Comment by neocarmack** [comment]
Becasue this two options offer the same security level, but using a public/private key pair instead of R and Hash(R) makes the source code looks consistent and concise. We may just simply apply elliptic curve to generate the temp pub/priv key pair instead of applying a complex HD chain, when creating the multi-sig address. 

And custom redeem script, which uses Hash(R) to lock an omnilayer transaction, currently not officially be supported by omnicore v0.18, although there is a way to contruct such a redeem script. 

OmniLayer is in a big update that is going to support SegWit. When the update releases, we may be able to invoke official interface to constuct transaction by custom redeem script. And at that time, there is a concise way to simply use a secret key R. 

Both of the two options are same to us.

**Comment by sidhujag** [comment]
> Becasue this two options offer the same security level, but using a public/private key pair instead of R and Hash(R) makes the source code looks consistent and concise. We may just simply apply elliptic curve to generate the temp pub/priv key pair instead of applying a complex HD chain, when creating the multi-sig address.
> 
> And custom redeem script, which uses Hash(R) to lock an omnilayer transaction, currently not officially be supported by omnicore v0.18, although there is a way to contruct such a redeem script.
> 
> OmniLayer is in a big update that is going to support SegWit. When the update releases, we may be able to invoke official interface to constuct transaction by custom redeem script. And at that time, there is a concise way to simply use a secret key R.
> 
> Both of the two options are same to us.

oh ok so you were constrained with not being able to use custom redeem scripts in omnicore (although its based on bitcoin? should be no problem) but ok, now it makes sense.

So once you support segwit you can do custom redeem script and not just a standard 2 of 2 multisig.

**Comment by sidhujag** [comment]
Could you use SIGHASH_NOINPUT and just exchange signatures of the tx prior to broadcast?

**Comment by neocarmack** [comment]
>Could you use SIGHASH_NOINPUT and just exchange signatures of the tx prior to broadcast?

Not yet. After we figure out the real benefit of SIGHASH_NOINPUT, we shall propose to Omni Layer Enhancement (proposal) for tech enhancement.

Currently we simply exchange signatures between peers when contruct transactions, and it works as of now.

There was a proposal to improve the punishment mechanism, called Eltoo, that requires the support of SIGHASH_NOINPUT. But at this stage, i dont see the benifits of Eltoo. Let me think about it after OmniBOLT testnet is online.

**Comment by sidhujag** [comment]
SIGHASH_NOINPUT is part of the official LN paper, and is required in order to work around issues where commitment transactions required non-malleable way to build off of a SIGHASH_NOINPUT transactions without affecting the transaction hash, so once a final commitment is created it can be published along with the parent once parent confirms and is buried deep enough.

**Comment by neocarmack** [comment]
Your are right.  SIGHASH_NOINPUT is part of the whitepaper. But we dont follow all the details of the whitepaper, we just follow the principal of the whitepaper. Especially in the context of Omnilayer, the funding process is a little more complex:

Creating an omni transaction requires a little BTC as miner fee, otherwise it will fail even if it is  unboardcast. So first of all, when we create a channel, we have to fund the channel a little BTC, say 0.5 usd value of BTC, as the miner fee. This is the cost of building a channel, and this is not revockable. 

https://api.omnilab.online/#fundingbtc

After that, we are able to create, say 10000 USDT, funding transaction but not broadcast before commitment tx C1a, revockable delivery tx RD1a, and breach remedy tx BR1a(maybe later) are correctly created in the channel, and signed by both participants. At this time, the channel has some real BTC(0.5 usd) in it, and is waiting for the 10000 USDT funding tx to be broadcasted.

These internal transactions are based on the BTC funding transaction as the parent. So this the reason why we dont use SIGHASH_NOINPUT. 

BTW, is SIGHASH_NOINPUT officially enbled by btc-core?

**Comment by sidhujag** [comment]
> Your are right. SIGHASH_NOINPUT is part of the whitepaper. But we dont follow all the details of the whitepaper, we just follow its principal. Especially in the context of Omnilayer, the funding process is a little more complex:
> 
> Creating an omni transaction requires a little BTC as miner fee, otherwise it will fail even if it is unboardcast. So first of all, when we create a channel, we have to fund the channel a little BTC, say 0.5 usd value of BTC, as the miner fee. This is the cost of building a channel, and this is not revockable.
> 
> https://api.omnilab.online/#fundingbtc
> 
> After that, we are able to create, say 10000 USDT, funding transaction but not broadcast before commitment tx C1a, revockable delivery tx RD1a, and breach remedy tx BR1a(maybe later) are correctly created in the channel, and signed by both participants. At this time, the channel has some real BTC(0.5 usd) in it, and is waiting for the 10000 USDT funding tx to be broadcasted.
> 
> These internal transactions are based on the BTC funding transaction as the parent. So this the reason why we dont use SIGHASH_NOINPUT.
> 
> BTW, is SIGHASH_NOINPUT officially enbled by btc-core?

yes that is fine, you are creating multisig and funding it with btc for gas and then the omni tx is funding with asset but not broadcast so it should use sighash_noinput and you can follow the same process as LN. Yes it is enabled along with segwit soft fork. You can also just send the btc in another output alongside the opreturn for omni tx, so you can put gas in and asset in the same tx, unbroadcasted.

**Comment by sidhujag** [comment]
@neocarmack do you use private keys in the commitment txs because omni is account based, or has omni changed to utxo input/output based now? From what I understand the commitment txs would take the input from a previous commitment and create a new CLTV output, and if the asset model is not utxo based that poses problems. Does using private key alleviate these concerns somehow?

**Comment by neocarmack** [comment]
>Why not use the same transaction to send the btc and the asset (which is in the opreturn output)....

Omnicore currently can not send both BTC and token in one transaction, although when sending tokens, some BTC "dust" will be sent to the destinine address.

**Comment by neocarmack** [comment]
> do you use private keys in the commitment txs because omni is account based, or has omni changed to utxo input/output based now? From what I understand the commitment txs would take the input from a previous commitment and create a new CLTV output, and if the asset model is not utxo based that poses problems with the BOLT spec, because you can wait for timeout and just use another output, or some historical output in the same address for spend-ability of that asset. 

I do not exactly understand your point. We do use private keys and use the counterparty's signature as well, in contructing commitment transacitons. 

And in contructing the following transactions, like breach remedy tx, htlc timeout tx, htlc execution delivery tx, etc, all these need private keys/signature in contructing. In our implementation of OBD, the commitment tx itself takes the funding transaction as input, not the previous commitment transaciton. So all the commitment txs are not chained, but the sons of one funding transaction.

**Comment by sidhujag** [comment]
what prevents someone from putting an older commitment on the chain? You use account model or utxo model for your asset protocol?

If account model, also what stops using the perpetrator from funding the asset to move based on another input (proving you are owner of the address but not the specific UTXO that is under lock.

**Comment by neocarmack** [comment]
>Also note that the official spec now has support for generic TLV's so you can put your asset data into there and perhaps just follow original spec.

before we develop this omni-asset-aware version of lighnting network, we consulted the lnd developer team if lnd can support smart asset or have a plan on this. The answer is no, because of the huge cost/effort that will be paid. Refactoring the code base to support smart asset is terrible.

After OmniBOLT network is online, we will use atomic swap to communicat lnd network.

**Comment by neocarmack** [comment]
>what prevents someone from putting an older commitment on the chain? You use account model or utxo model for your asset protocol?

The breach remedy(RD) transaction prevents broadcasting older commitmen transaction.

**Comment by neocarmack** [comment]
BTW, we know RGB, which still uses liquidity of BTC in lnd network. Omni team had cummunication with RGB team last year.  Wish they have good luck.

OmniBOLT testnet will be online soon. At that time, we will connect to LND network. We both has advantages: OmniBOLT has liquidity of assets, including stable coins, while LND has a good liquidity of BTC.

**Comment by sidhujag** [comment]
> > what prevents someone from putting an older commitment on the chain? You use account model or utxo model for your asset protocol?
> 
> The breach remedy(RD) transaction prevents broadcasting older commitmen transaction.

Ok I am just not sure how that locking to an address would work when the commitment CLTV transactions are clearly locking UTXOs, I just feel there is a security vulnerability there but I am not sure what that is at the moment, Rusty Russel was sure that it would be an issue if assets are not locked to UTXOs, Colu seems to be doing so and there was a comment on there if Omni was changing to utxo and it looks like RGB is also utxo based I think.

**Comment by sidhujag** [comment]
> > Also note that the official spec now has support for generic TLV's so you can put your asset data into there and perhaps just follow original spec.
> 
> before we develop this omni-asset-aware version of lighnting network, we consulted the lnd developer team if lnd can support smart asset or have a plan on this. The answer is no, because of the huge cost/effort that will be paid. Refactoring the code base to support smart asset is terrible.
> 
> After OmniBOLT network is online, we will use atomic swap to communicat lnd network.

Its not a refactor, TLV are an extension for general purpose functionality enhancement.

**Comment by neocarmack** [comment]
>Its not a refactor, TLV are an extension for general purpose functionality enhancement. They are already used internally for tx data, but an extended field was added for external data

Any reference link available for this enhancement?

**Comment by neocarmack** [comment]
>Its not a refactor, TLV are an extension for general purpose functionality enhancement. They are already used internally for tx data, but an extended field was added for external data

I'm not sure if TLV filed can be used to put colored coin information in. What happens when broadcasting transaction to withdraw coin defined in TLV back to BTC network(address)? Do we really need to re-invent another omnilayer to support it using OP_RETURN again? I dont see necessity of doing so.

Lighning network itself has no temper-resistent mechanism. What if someone modifies the data of TLV, and there is no protective protocol like omnilayer to prevent broadcasting dirty data?

I remember before we decide to develop OmniBOLT, we studied and consulted lnd team, about the custom field for adding omni asset data in. But no positive answer at that time. 

I guess i need some time to figure it out. Some of these problems seems unsolvable.

**Comment by sidhujag** [comment]
It’s just a spec, every node I assume verifies the transaction including opreturn data to color the outputs. What it is about is using the official spec whilst encoding data in tlv so it can be quickly filtered. That is based on my intuition though. Rgb guys seemed pretty adamant that the tlv extension capability was needed so it can widely support multi coin ln networks.. maybe their transaction is not standard, but if every tx is parsed and verified prior to relay to another node you are right you shouldn’t need extra info. But then I’m not sure you even need asset Id field in there either? You can color that locally once you parse txs if you need to filter txs by asset.

**Comment by neocarmack** [comment]
So if RBG heavily rely on TLV extention, then they shall put more effort on the other end of the whole workflow: the consensus layer of colored coin, where currently solved by omnilayer/btc, or similar protocols. Because on lighnting network itself, no more works can be done.

If one node involved in a hop of transferring money, it shall verify the data, but how? There is no chain, no consensus algorithm, can be introduced into the process. Even watch tower, i dont think it is a solution. The only check point is when broadcasting, let POW to check. So that's where omnilayer waiting for.

Lots of doubtable questions on that technical path...............

Maybe i did not fully understand your point..........But its great to have chat with you.

**Comment by sidhujag** [comment]
> > Its not a refactor, TLV are an extension for general purpose functionality enhancement. They are already used internally for tx data, but an extended field was added for external data
> 
> Any reference link available for this enhancement? except the official RFC:
> 
> https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md#type-length-value-format

https://github.com/lightningnetwork/lightning-rfc/pull/754 which is merged recently and referenced from https://github.com/lightningnetwork/lightning-rfc/pull/72

Great to chat with you aswell, I am working in the similar field so that is why I have questions, hopefully with mindshare we can solve these problems and bring this technology to the world in meaningful ways.

**Comment by sidhujag** [comment]
> So if RBG heavily rely on TLV extention, then they shall put more effort on the other end of the whole workflow: the consensus layer of colored coin, where currently solved by omnilayer/btc, or similar protocols. Because on lighnting network itself, no more works can be done.
> 
> If one node involved in a hop of transferring money, it shall verify the data, but how? There is no chain, no consensus algorithm, can be introduced into the process. Even watch tower, i dont think it is a solution. The only check point is when broadcasting, let POW to check. So that's where omnilayer waiting for.
> 
> Lots of doubtable questions on that technical path...............
> 
> Maybe i did not fully understand your point..........But its great to have chat with you.

I meant that on each hop the transaction in question can be parsed to get the colouring information regarding the asset transfer and if accepted (if their channel is of the same asset and has liquidity then hop is valid)

**Comment by sidhujag** [comment]
@neocarmack See my response on https://github.com/lightningnetwork/lightning-rfc/pull/72 about using LN with multiple-assets.. it only makes sense for single asset networks.. hopefully you have designed with it in mind (that LN network only work on single assets at a time)

**Comment by neocarmack** [comment]
>Edit: I might retract this, because the wire protocol is different it does not relay the entire transaction but amount, payment hash and hop information. I guess that is why you have asset id as it is passed along on wire and used to detect which other channels are open with that asset as well. Maybe with TLV extension you can sign a payload which will allow one to verify it when recreating the commitment tx locally.

Emmm, we will try TLV extention to embed OmniBOLT tx info into a general lnd message, so that lnd  network will have the liquidity of omnilayer assets. 

But the top priority on our schedule is release our first omnibolt testnet, on May 30.

**Comment by neocarmack** [comment]
>@neocarmack See my response on lightningnetwork/lightning-rfc#72 about using LN with multiple-assets.. it only makes sense for single asset networks.. hopefully you have designed with it in mind (that LN network only work on single assets at a time)

It's true, LN channel only work on one asset at a time. We have not finalize the design yet. What we focus is how to communicate with existing lnd network, not just embed omnibolt transaction into lnd messages. 

I dont know which one will be our best choice.

**Comment by sidhujag** [comment]
> > @neocarmack See my response on [lightningnetwork/lightning-rfc#72](https://github.com/lightningnetwork/lightning-rfc/pull/72) about using LN with multiple-assets.. it only makes sense for single asset networks.. hopefully you have designed with it in mind (that LN network only work on single assets at a time)
> 
> It's true, LN channel only work on one asset at a time. We have not finalize the design yet. What we focus is how to communicate with existing lnd network, not just embed omnibolt transaction into lnd messages.
> 
> I dont know which one will be our best choice.

You mean some nodes will propagate and maybe even participate not knowing or caring that it’s an omni asset

**Comment by neocarmack** [comment]
>You mean some nodes will propagate and maybe even participate not knowing or caring that it’s an omni asset

Correct. Embedding info into TLV field just like sending message via some instant message systems. LND has no knowledge of any command contained in TLV.

I consulted lnd community, the answer is that we can subscribe invoice to monitor the event of payment ( which is still unclear to me). But this is not what we want. 

I put the technology guild in this repo, hope this can help you see the whole picture of OmniBOLT, especially what we are currently trying to communicate with LND:

https://github.com/omnilaboratory/OmniBOLT-spec/tree/master/docs
